src/example/RouterImmutable.sol

```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: MIT
+// @author: thirdweb (https://github.com/thirdweb-dev/dynamic-contracts)
+
+pragma solidity ^0.8.0;
+
+import "../presets/BaseRouterWithDefaults.sol";
+
+/**
+ *  This smart contract is an EXAMPLE, and is not meant for use in production.
+ */
+
+contract RouterImmutable is BaseRouterWithDefaults {
+    
+    constructor(Extension[] memory _extensions) BaseRouterWithDefaults(_extensions) {}
+
+    /*///////////////////////////////////////////////////////////////
+                            Overrides
+    //////////////////////////////////////////////////////////////*/
+
+    /// @dev Returns whether a new extension can be added in the given execution context.
+    function _canAddExtension(Extension memory) internal virtual override returns (bool) {
+        return false;
+    }
+
+    /// @dev Returns whether an extension can be replaced in the given execution context.
+    function _canReplaceExtension(Extension memory) internal view virtual override returns (bool) {
+        return false;
+    }
+
+    /// @dev Returns whether an extension can be removed in the given execution context.
+    function _canRemoveExtension(string memory) internal virtual override returns (bool) {
+        return false;
+    }
+
+    /// @dev Returns whether a function can be added to an extension in the given execution context.
+    function _canAddFunctionToExtension(string memory, ExtensionFunction memory) internal view virtual override returns (bool) {
+        return false;
+    }
+
+    /// @dev Returns whether an extension can be removed from an extension in the given execution context.
+    function _canRemoveFunctionFromExtension(string memory, bytes4) internal view virtual override returns (bool) {
```

Comment:
do we  really need a permission function for each action?

Reply:
I've written `_can` functions for each function and placed each function's relevant require checks in them. I like the pattern, and doesn't limit the `_can` functions to permissions. What do you think?

Reply:
it just feels heavy to have to override all these _can functions when most of the time you'd have the same logic in them no?

Reply:
Makes sense. Added an `isAuthorizedCallToUpgrade` function to `ExtensionManager`. You can check the `/example` contracts for the usage of `isAuthorizedCallToUpgrade` and `_can...` functions.